steering_controllers_library/src/steering_odometry.cpp

```diff
@@ -0,0 +1,338 @@
+}
+
+void SteeringOdometry::set_wheel_params(double wheel_radius, double wheelbase, double wheel_track)
+{
+  wheel_radius_ = wheel_radius;
+  wheelbase_ = wheelbase;
+  wheel_track_ = wheel_track;
+}
+
+void SteeringOdometry::set_velocity_rolling_window_size(size_t velocity_rolling_window_size)
+{
+  velocity_rolling_window_size_ = velocity_rolling_window_size;
+
+  reset_accumulators();
+}
+
+void SteeringOdometry::set_odometry_type(const unsigned int type) { config_type_ = type; }
+
+double SteeringOdometry::convert_trans_rot_vel_to_steering_angle(double Vx, double theta_dot)
+{
+  if (theta_dot == 0 || Vx == 0)
+  {
+    return 0;
+  }
+  return std::atan(theta_dot * wheelbase_ / Vx);
+}
+
+std::tuple<std::vector<double>, std::vector<double>> SteeringOdometry::get_commands(
+  double Vx, double theta_dot)
+{
+  // desired velocty and steering angle of the midle of traction and steering axis
+  double Ws, alpha;
+
+  if (Vx == 0 && theta_dot != 0)
+  {
+    alpha = theta_dot > 0 ? M_PI_2 : -M_PI_2;
+    Ws = abs(theta_dot) * wheelbase_ / wheel_radius_;
+  }
+  else
+  {
+    alpha = SteeringOdometry::convert_trans_rot_vel_to_steering_angle(Vx, theta_dot);
+    Ws = Vx / (wheel_radius_ * std::cos(steer_pos_));
+  }
+
+  if (config_type_ == BICYCLE_CONFIG)
+  {
+    std::vector<double> traction_commands = {Ws};
+    std::vector<double> steering_commands = {alpha};
+    return std::make_tuple(traction_commands, steering_commands);
+  }
+  else if (config_type_ == TRICYCLE_CONFIG)
+  {
+    std::vector<double> traction_commands;
+    std::vector<double> steering_commands;
+    if (fabs(steer_pos_) < 1e-6)
+    {
+      traction_commands = {Ws, Ws};
+    }
+    else
+    {
+      double turning_radius = wheelbase_ / std::tan(steer_pos_);
+      double Wr = Ws * (turning_radius + wheel_track_ * 0.5) / turning_radius;
+      double Wl = Ws * (turning_radius - wheel_track_ * 0.5) / turning_radius;
+      traction_commands = {Wr, Wl};
+    }
+    steering_commands = {alpha};
+    return std::make_tuple(traction_commands, steering_commands);
+  }
+  else if (config_type_ == ACKERMANN_CONFIG)
+  {
+    std::vector<double> traction_commands;
+    std::vector<double> steering_commands;
+    if (fabs(steer_pos_) < 1e-6)
+    {
+      traction_commands = {Ws, Ws};
+    }
+    else
+    {
+      double turning_radius = wheelbase_ / std::tan(steer_pos_);
+      double Wr = Ws * (turning_radius + wheel_track_ * 0.5) / turning_radius;
+      double Wl = Ws * (turning_radius - wheel_track_ * 0.5) / turning_radius;
+      traction_commands = {Wr, Wl};
+    }
+    // TODO(petkovich): implement this
+    // double wanted_turning_radius = wheelbase_ / std::tan(alpha);
+    double alpha_r = alpha;
+    double alpha_l = alpha;
+    steering_commands = {alpha_r, alpha_l};
+    return std::make_tuple(traction_commands, steering_commands);
+  }
+  else
+  {
+    throw std::runtime_error("Config not implemented");
```

Comment:
Is it necessary to throw exceptions that we don't catch at all in these scenarios? I might be wrong, but calling std::terminate seems a bit overkill to me when a service is unavailable or the config is missing.

From a quick search I've seen something similar in the diff drive controller's code though.

Reply:
The exception will be triggered only if the config_type_ isn't set to bicycle/tricycle/ackermann. The idea here was to ensure all non-canonical configurations implemented on top of the `steering_controllers_library` need to add and set their own `config_type_` if their fwd kinematics needs customization. Otherwise, they need to set `config_type_` to one of the three canonical configs. 
I'm leaning towards @destogl last idea because only returning NaN could lead the developer towards debug divisions and multiplications while the root cause is the unset parameter. So, the warning would be printed in the `steering_odometry.cpp` and I can return NaN tuple to avoid throwing exceptions?